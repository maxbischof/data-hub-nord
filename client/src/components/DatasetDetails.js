import React, { useState } from 'react'
import styled from 'styled-components'
import Table from './Table'
import PropTypes from 'prop-types'
import { useCSV } from '../hooks/useCSV'
import LoadingDots from './ui/LoadingDots'
import Map from './Map'
import MapForm from './MapForm'
import Button from './ui/Button'

export default function DatasetDetails({
  title,
  description,
  license,
  publisher,
  url,
  removeColumns,
  columnsOrder,
}) {
  const tableData = useCSV({
    url: url,
    removeColumns: removeColumns,
    columnsOrder: columnsOrder,
  })

  const columnNames = tableData && Object.keys(tableData[0])

  const [mapData, setMapData] = useState()
  const [showMapForm, setShowMapForm] = useState(false)

  return (
    <main>
      <DetailsDescription>
        <Headline>{title}</Headline>
        <Paragraph>{description}</Paragraph>
        <br />
        <Headline3>Lizenz</Headline3>
        <Paragraph>
          <a href={license}>Zur Lizenz</a>
        </Paragraph>
        <br />
        {publisher ? (
          <>
            <Headline3>Herausgeber</Headline3>
            <Paragraph>{publisher}</Paragraph>
          </>
        ) : (
          ''
        )}
      </DetailsDescription>

      {tableData ? (
        <VisualisationSection>
          <Headline2>Karte</Headline2>
          <Paragraph>
            Ist eine Adresse oder sind geographische Koordinaten in der Tabelle
            vorhanden? Dann erstelle eine Karte.
          </Paragraph>
          {!showMapForm && (
            <StyledButton styleType="more" onClick={() => setShowMapForm(true)}>
              Karte erstellen
            </StyledButton>
          )}
          {!mapData && showMapForm && (
            <MapForm
              tableData={tableData}
              columnNames={columnNames}
              setMapData={setMapData}
            />
          )}
          {mapData && <Map rows={mapData} columnNames={columnNames} />}
          <Headline2>Tabelle</Headline2>
          <Table data={tableData}></Table>
        </VisualisationSection>
      ) : (
        <LoadingDots />
      )}
    </main>
  )
}

DatasetDetails.propTypes = {
  title: PropTypes.string.isRequired,
  description: PropTypes.string,
  license: PropTypes.string.isRequired,
  publisher: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
  keys: PropTypes.array,
  seperator: PropTypes.string,
}

const StyledButton = styled(Button)`
  align-self: center;
  margin: 15px;
`

const DetailsDescription = styled.div`
  margin: 0 37px 30px 37px;
  max-width: 600px;
`
const VisualisationSection = styled.div`
  margin: 0 37px 30px 37px;
  display: flex;
  flex-direction: column;
`

const Headline = styled.h1`
  font-size: 25px;
  margin: 30px 0 10px 0;
  padding: 0;
`

const Headline2 = styled.h2`
  font-size: 22px;
  padding: 0;
  margin: 30px 37px 0 0;
`

const Headline3 = styled.h3`
  font-size: 17px;
  margin: 0;
  padding: 0;
`

const Paragraph = styled.p`
  margin: 0;
  padding: 0;
`
