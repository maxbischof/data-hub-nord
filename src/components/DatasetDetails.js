import React, { useState, useEffect } from "react"
import styled from "styled-components"
import Table from "./Table"
import PropTypes from "prop-types"
import { fetchCSV, csvToObjectsArray } from "../utils.js"

export default function DatasetDetails({ datasetDescription }) {
  const {
    imageUrl,
    name,
    description,
    license,
    publisher,
    url,
    keys,
    seperator,
  } = datasetDescription

  const [tableData, setTableData] = useState([])

  useEffect(() => {
    fetchCSV({ path: url }).then((response) => {
      const array = csvToObjectsArray({
        csv: response,
        columnNames: keys,
        seperator: seperator,
      })
      setTableData(array)
    })
  }, [keys, seperator, url])

  return (
    <main>
      <DetailsDescription>
        <DetailsPicture src={imageUrl} />
        <Headline>{name}</Headline>
        <Paragraph>{description}</Paragraph>
        <br />
        <Headline3>Lizenz</Headline3>
        <Paragraph>{license}</Paragraph>
        <br />
        <Headline3>Herausgeber</Headline3>
        <Paragraph>{publisher}</Paragraph>
      </DetailsDescription>

      <Headline2>Tabelle</Headline2>
      <Table tableData={tableData}></Table>
    </main>
  )
}

DatasetDetails.propTypes = {
  datasetDescription: PropTypes.object.isRequired,
}

const DetailsDescription = styled.div`
  margin: 0 37px 30px 37px;
`

const DetailsPicture = styled.img`
  border-radius: 5px;
  border: 1px solid var(--grey);
  width: 100%;
  margin: 0 0 30px 0;
`

const Headline = styled.h1`
  font-size: 25px;
  margin: 0 0 10px 0;
  padding: 0;
`

const Headline2 = styled.h2`
  font-size: 22px;
  padding: 0;
  margin: 30px 37px 0 37px;
`

const Headline3 = styled.h3`
  font-size: 17px;
  margin: 0;
  padding: 0;
`

const Paragraph = styled.p`
  margin: 0;
  padding: 0;
`
