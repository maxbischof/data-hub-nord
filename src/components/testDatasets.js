export const datasets = [
  {
    id: 1,
    name: "Arbeitslosigkeit Kiel",
    description: "Arbeitslosigkeit in Kiel absolut und relativ seit 2009.",
    url:
      "https://www.kiel.de/opendata/kiel_wirtschaft_arbeit_unterbeschaeftigung.csv",
    imageUrl:
      "https://images.unsplash.com/photo-1529400971008-f566de0e6dfc?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=800&q=80",
    keys: [],
    license: "Datenlizenz Deutschland – Zero – Version 2.0",
    publisher: "Landeshauptstadt Kiel",
    displayedAs: "",
    seperator: ";"
  },
  {
    id: 2,
    name: "Badegewässer Messungen",
    description: "Messungen der Badegewässerqualität in Schleswig Holstein.",
    url:
      "http://efi2.schleswig-holstein.de/bg/opendata/v_proben_odata.csv",
    imageUrl:
      "https://images.unsplash.com/photo-1527694545342-6176693ca513?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=800&q=60",
    keys: [
      'BADEGEWAESSERID',
      'MESSSTELLENNAME',
      'MESSSTELLENID',
      'UEBERWACHUNGSARTID',
      'UEBERWASCHUNGSARTTEXT',
      'GEWAESSERKATEGORIE',
      'KUESTENGEWAESSER',
      'PROBEID',
      'DATUMMESSUNG',
      'PROBENART',
      'ECOLI',
      'INTEST_ENTEROKOKKEN',
      'WASSERTEMP',
      'LUFTTEMP',
      'SICHTTIEFE',
      'BEMERKUNG'],
    license: "Datenlizenz Deutschland – Zero – Version 2.0",
    publisher: "Landeshauptstadt Kiel",
    displayedAs: "",
    seperator: "|"
  },
]
