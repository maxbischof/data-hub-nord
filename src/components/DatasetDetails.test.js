import React from "react"
import DatasetDetails from "./DatasetDetails"
import renderer from "react-test-renderer"
import { datasets } from "./testDatasets.js"

it("renders datasetdetails component", () => {
  const renderedDatasetDetails = renderer
    .create(<DatasetDetails {...datasets[0]}/>)
    .toJSON()
  expect(renderedDatasetDetails).toMatchSnapshot()
})
