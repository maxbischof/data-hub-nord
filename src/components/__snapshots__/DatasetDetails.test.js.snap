// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders datasetdetails component 1`] = `
<main>
  <div
    className="sc-AxgMl jzbxKt"
  >
    <img
      className="sc-AxheI fKDPmu"
      src="https://images.unsplash.com/photo-1529400971008-f566de0e6dfc?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1950&q=80"
    />
    <h1
      className="sc-Axmtr fZZTNn"
    >
      Arbeitslosigkeit Kiel
    </h1>
    <p
      className="sc-fzoLsD kLkXrb"
    >
      Arbeitslosigkeit in Kiel absolut und relativ seit 2009
    </p>
    <br />
    <h3
      className="sc-fzozJi keEFn"
    >
      Lizenz
    </h3>
    <p
      className="sc-fzoLsD kLkXrb"
    >
      Datenlizenz Deutschland – Zero – Version 2.0
    </p>
    <br />
    <h3
      className="sc-fzozJi keEFn"
    >
      Herausgeber
    </h3>
    <p
      className="sc-fzoLsD kLkXrb"
    >
      Landeshauptstadt Kiel
    </p>
  </div>
  <h2
    className="sc-AxmLO eCGjlW"
  >
    Tabelle
  </h2>
  <div
    className="sc-AxjAm bSmOPU"
  >
    <div
      className="sc-AxhCb eHOKVW"
    />
    <table
      className="sc-AxirZ ijqXlW"
    >
      <thead
        className="sc-AxiKw cObrap"
      >
        <tr />
      </thead>
      <tbody
        className="sc-AxhUy bLwAyT"
      >
        <tr />
      </tbody>
    </table>
  </div>
</main>
`;
